File: warehouse-backend/personnel/sheba_utils.py
"""
Iranian IBAN (Sheba) Validation and Bank Directory Utility
Standard: ISO 7064 Mod 97-10 & ISO 13616
"""
import re
from typing import Optional, Dict, Tuple

# لیست کامل ۳۲ بانک و مؤسسه اعتباری رسمی کشور با کد ۳ رقمی استاندارد شبا (CBI Official Directory)
IRANIAN_BANKS_MAP: Dict[str, Dict[str, str]] = {
    '010': {'name': 'بانک مرکزی جمهوری اسلامی ایران', 'short_name': 'مرکزی', 'icon': '🏛️'},
    '011': {'name': 'بانک صنعت و معدن', 'short_name': 'صنعت و معدن', 'icon': '🏭'},
    '012': {'name': 'بانک ملت', 'short_name': 'ملت', 'icon': '🔴'},
    '013': {'name': 'بانک رفاه کارگران', 'short_name': 'رفاه', 'icon': '🔵'},
    '014': {'name': 'بانک مسکن', 'short_name': 'مسکن', 'icon': '🏠'},
    '015': {'name': 'بانک سپه', 'short_name': 'سپه', 'icon': '⭐'},
    '016': {'name': 'بانک کشاورزی', 'short_name': 'کشاورزی', 'icon': '🌾'},
    '017': {'name': 'بانک ملی ایران', 'short_name': 'ملی', 'icon': '🦁'},
    '018': {'name': 'بانک تجارت', 'short_name': 'تجارت', 'icon': '🔷'},
    '019': {'name': 'بانک صادرات ایران', 'short_name': 'صادرات', 'icon': '💎'},
    '020': {'name': 'بانک توسعه صادرات ایران', 'short_name': 'توسعه صادرات', 'icon': '🚢'},
    '021': {'name': 'پست بانک ایران', 'short_name': 'پست بانک', 'icon': '📮'},
    '022': {'name': 'بانک توسعه تعاون', 'short_name': 'توسعه تعاون', 'icon': '🤝'},
    '051': {'name': 'موسسه اعتباری غیربانکی ملل (عسکریه)', 'short_name': 'ملل', 'icon': '🏢'},
    '053': {'name': 'بانک کارآفرین', 'short_name': 'کارآفرین', 'icon': '💼'},
    '054': {'name': 'بانک پارسیان', 'short_name': 'پارسیان', 'icon': '🌸'},
    '055': {'name': 'بانک اقتصاد نوین', 'short_name': 'اقتصاد نوین', 'icon': '🌐'},
    '056': {'name': 'بانک سامان', 'short_name': 'سامان', 'icon': '🌊'},
    '057': {'name': 'بانک پاسارگاد', 'short_name': 'پاسارگاد', 'icon': '🏛️'},
    '058': {'name': 'بانک سرمایه', 'short_name': 'سرمایه', 'icon': '📊'},
    '059': {'name': 'بانک سینا', 'short_name': 'سینا', 'icon': '💠'},
    '060': {'name': 'بانک قرض‌الحسنه مهر ایران', 'short_name': 'مهر ایران', 'icon': '☀️'},
    '061': {'name': 'بانک شهر', 'short_name': 'شهر', 'icon': '🏙️'},
    '062': {'name': 'بانک آینده', 'short_name': 'آینده', 'icon': '🔮'},
    '063': {'name': 'بانک انصار (ادغام در سپه)', 'short_name': 'انصار', 'icon': '⭐'},
    '064': {'name': 'بانک گردشگری', 'short_name': 'گردشگری', 'icon': '🧭'},
    '065': {'name': 'بانک حکمت ایرانیان (ادغام در سپه)', 'short_name': 'حکمت', 'icon': '⭐'},
    '066': {'name': 'بانک دی', 'short_name': 'دی', 'icon': '🦅'},
    '069': {'name': 'بانک ایران زمین', 'short_name': 'ایران زمین', 'icon': '🌍'},
    '070': {'name': 'بانک قرض‌الحسنه رسالت', 'short_name': 'رسالت', 'icon': '🌿'},
    '075': {'name': 'موسسه اعتباری غیربانکی نور', 'short_name': 'نور', 'icon': '💡'},
    '078': {'name': 'بانک خاورمیانه', 'short_name': 'خاورمیانه', 'icon': '🗺️'},
}

def extract_sheba_digits(raw_input: str) -> str:
    """
    استخراج هوشمند صرفاً ارقام عددی از هر متن آلوده به حروف، فواصل و نشانه‌ها
    """
    if not raw_input:
        return ""
    persian_digits = "۰۱۲۳۴۵۶۷۸۹"
    arabic_digits = "٠١٢٣٤٥٦٧٨٩"
    cleaned = str(raw_input)
    for i, d in enumerate(persian_digits):
        cleaned = cleaned.replace(d, str(i))
    for i, d in enumerate(arabic_digits):
        cleaned = cleaned.replace(d, str(i))
    
    digits_only = re.sub(r'\D', '', cleaned)
    return digits_only[:24]

def extract_account_number_from_sheba(raw_sheba: str) -> str:
    """
    استخراج شماره حساب بانکی از داخل ارقام شماره شبا (حذف صفرهای زائد سمت چپ)
    """
    digits = extract_sheba_digits(raw_sheba)
    if len(digits) <= 6:
        return ""
    raw_account = digits[6:]
    trimmed = raw_account.lstrip('0')
    return trimmed or raw_account

def get_bank_by_name(bank_name: str) -> Optional[Tuple[str, Dict[str, str]]]:
    """
    یافتن کد و مشخصات بانک بر اساس نام
    """
    if not bank_name:
        return None
    name_clean = bank_name.strip()
    for code, info in IRANIAN_BANKS_MAP.items():
        if info['name'] == name_clean or info['short_name'] == name_clean or code == name_clean:
            return code, info
        if name_clean in info['name'] or info['name'] in name_clean:
            return code, info
    return None

def validate_account_number(account_number: str) -> Tuple[bool, Optional[str], str]:
    """
    اعتبارسنجی سخت‌گیرانه شماره حساب بانکی بر اساس قوانین بانکی کشور
    خروجی: (is_valid, error_message, clean_account_number)
    """
    if not account_number or not str(account_number).strip():
        return False, "شماره حساب نمی‌تواند خالی باشد.", ""
    
    clean_digits = extract_sheba_digits(str(account_number))
    if not clean_digits:
        return False, "شماره حساب فاقد هرگونه رقم عددی است.", ""
    
    if re.match(r'^0+$', clean_digits):
        return False, "شماره حساب نمی‌تواند صفر یا تماماً صفر باشد.", clean_digits

    if len(clean_digits) < 4:
        return False, f"طول شماره حساب ناقص است ({len(clean_digits)} از حداقل ۴ رقم).", clean_digits

    if len(clean_digits) > 18:
        return False, "طول شماره حساب بیشتر از سقف مجاز بانکی (۱۸ رقم) است.", clean_digits

    return True, None, clean_digits

def generate_sheba_from_account(bank_identifier: str, account_number: str, account_type: str = "0") -> str:
    """
    تولید خودکار شماره شبا بر مبنای استاندارد ISO 7064 Mod 97-10 از روی شماره حساب و بانک با گارد اعتبارسنجی
    """
    if not bank_identifier or not account_number:
        return ""
    
    is_valid, _, clean_acc = validate_account_number(account_number)
    if not is_valid:
        return ""

    bank_code = None
    if bank_identifier in IRANIAN_BANKS_MAP:
        bank_code = bank_identifier
    else:
        bank_lookup = get_bank_by_name(bank_identifier)
        if bank_lookup:
            bank_code = bank_lookup[0]

    if not bank_code:
        return ""

    padded_account = clean_acc.zfill(18)
    bban = f"{bank_code}{account_type or '0'}{padded_account}"
    payload = f"{bban}182700"

    try:
        rem = int(payload) % 97
        check_digits = str(98 - rem).zfill(2)
        return f"IR{check_digits}{bban}"
    except Exception:
        return ""

def clean_sheba(raw_sheba: str) -> str:
    """
    نرمال‌سازی و پاکسازی کاراکترهای شماره شبا و افزودن خودکار IR
    """
    if not raw_sheba:
        return ""
    digits = extract_sheba_digits(raw_sheba)
    if not digits:
        return ""
    return "IR" + digits

def get_bank_from_sheba(raw_sheba: str) -> Optional[Dict[str, str]]:
    """
    استخراج مشخصات بانک از روی کد ۳ رقمی شبا
    """
    digits = extract_sheba_digits(raw_sheba)
    if len(digits) >= 5:
        bank_code = digits[2:5]
        return IRANIAN_BANKS_MAP.get(bank_code)
    return None

def validate_sheba(raw_sheba: str) -> Tuple[bool, Optional[str], Optional[Dict[str, str]]]:
    """
    اعتبارسنجی ریاضی شماره شبا بر مبنای استاندارد ISO 7064 Mod 97-10
    خروجی: (is_valid, error_message, bank_info)
    """
    if not raw_sheba or not str(raw_sheba).strip():
        return False, "شماره شبا نمی‌تواند خالی باشد.", None

    digits = extract_sheba_digits(raw_sheba)
    bank_info = get_bank_from_sheba(digits)

    if len(digits) == 0:
        return False, "شماره شبا فاقد هرگونه رقم عددی است.", None

    if len(digits) < 24:
        return False, f"طول ارقام شبا ناقص است ({len(digits)} از ۲۴ رقم).", bank_info

    check_digits = digits[:2]
    if check_digits in ['00', '01', '99']:
        return False, "ارقام کنترلی شبا نامعتبر است.", bank_info

    account_part = extract_account_number_from_sheba(digits)
    if not account_part or re.match(r'^0+$', account_part):
        return False, "شماره حساب مندرج در شبا نامعتبر (صفر) است.", bank_info

    bban = digits[2:]
    numeric_str = bban + "1827" + check_digits

    try:
        remainder = int(numeric_str) % 97
        if remainder == 1:
            return True, None, bank_info
        else:
            return False, "کد کنترلی شبا نامعتبر است (خطای محاسباتی Mod 97).", bank_info
    except Exception as e:
        return False, f"خطا در اعتبارسنجی ساختار شبا: {e}", bank_info

File: warehouse-front/src/app/core/utils/sheba-utils.ts
/**
 * Iranian IBAN (Sheba) Validation and Bank Detection Utility
 * Standard: ISO 7064 Mod 97-10 & ISO 13616
 */

export interface IranianBankInfo {
  code: string;       // کد ۳ رقمی شبا
  binCode?: string;   // پیش‌شماره ۶ رقمی کارت شتاب
  name: string;       // نام فارسی بانک
  shortName: string;  // نام کوتاه
  icon?: string;      // ایموجی یا آیکون
}

// لیست کامل ۳۲ بانک و مؤسسه اعتباری رسمی کشور با کد ۳ رقمی استاندارد شبا (CBI Official Directory)
export const IRANIAN_BANKS: IranianBankInfo[] = [
  { code: '010', binCode: '636795', name: 'بانک مرکزی جمهوری اسلامی ایران', shortName: 'مرکزی', icon: '🏛️' },
  { code: '011', binCode: '627760', name: 'بانک صنعت و معدن', shortName: 'صنعت و معدن', icon: '🏭' },
  { code: '012', binCode: '610433', name: 'بانک ملت', shortName: 'ملت', icon: '🔴' },
  { code: '013', binCode: '589463', name: 'بانک رفاه کارگران', shortName: 'رفاه', icon: '🔵' },
  { code: '014', binCode: '628023', name: 'بانک مسکن', shortName: 'مسکن', icon: '🏠' },
  { code: '015', binCode: '589210', name: 'بانک سپه', shortName: 'سپه', icon: '⭐' },
  { code: '016', binCode: '603770', name: 'بانک کشاورزی', shortName: 'کشاورزی', icon: '🌾' },
  { code: '017', binCode: '603799', name: 'بانک ملی ایران', shortName: 'ملی', icon: '🦁' },
  { code: '018', binCode: '627353', name: 'بانک تجارت', shortName: 'تجارت', icon: '🔷' },
  { code: '019', binCode: '603769', name: 'بانک صادرات ایران', shortName: 'صادرات', icon: '💎' },
  { code: '020', binCode: '627961', name: 'بانک توسعه صادرات ایران', shortName: 'توسعه صادرات', icon: '🚢' },
  { code: '021', binCode: '627648', name: 'پست بانک ایران', shortName: 'پست بانک', icon: '📮' },
  { code: '022', binCode: '502908', name: 'بانک توسعه تعاون', shortName: 'توسعه تعاون', icon: '🤝' },
  { code: '051', binCode: '621986', name: 'موسسه اعتباری غیربانکی ملل (عسکریه)', shortName: 'ملل', icon: '🏢' },
  { code: '053', binCode: '627412', name: 'بانک کارآفرین', shortName: 'کارآفرین', icon: '💼' },
  { code: '054', binCode: '622106', name: 'بانک پارسیان', shortName: 'پارسیان', icon: '🌸' },
  { code: '055', binCode: '627488', name: 'بانک اقتصاد نوین', shortName: 'اقتصاد نوین', icon: '🌐' },
  { code: '056', binCode: '621986', name: 'بانک سامان', shortName: 'سامان', icon: '🌊' },
  { code: '057', binCode: '502229', name: 'بانک پاسارگاد', shortName: 'پاسارگاد', icon: '🏛️' },
  { code: '058', binCode: '639607', name: 'بانک سرمایه', shortName: 'سرمایه', icon: '📊' },
  { code: '059', binCode: '639346', name: 'بانک سینا', shortName: 'سینا', icon: '💠' },
  { code: '060', binCode: '627381', name: 'بانک قرض‌الحسنه مهر ایران', shortName: 'مهر ایران', icon: '☀️' },
  { code: '061', binCode: '504706', name: 'بانک شهر', shortName: 'شهر', icon: '🏙️' },
  { code: '062', binCode: '636214', name: 'بانک آینده', shortName: 'آینده', icon: '🔮' },
  { code: '063', binCode: '627353', name: 'بانک انصار (ادغام در سپه)', shortName: 'انصار', icon: '⭐' },
  { code: '064', binCode: '505785', name: 'بانک گردشگری', shortName: 'گردشگری', icon: '🧭' },
  { code: '065', binCode: '636949', name: 'بانک حکمت ایرانیان (ادغام در سپه)', shortName: 'حکمت', icon: '⭐' },
  { code: '066', binCode: '505416', name: 'بانک دی', shortName: 'دی', icon: '🦅' },
  { code: '069', binCode: '505801', name: 'بانک ایران زمین', shortName: 'ایران زمین', icon: '🌍' },
  { code: '070', binCode: '504172', name: 'بانک قرض‌الحسنه رسالت', shortName: 'رسالت', icon: '🌿' },
  { code: '075', binCode: '606373', name: 'موسسه اعتباری غیربانکی نور', shortName: 'نور', icon: '💡' },
  { code: '078', binCode: '507677', name: 'بانک خاورمیانه', shortName: 'خاورمیانه', icon: '🗺️' }
];

export interface ShebaValidationResult {
  isValid: boolean;
  rawSheba: string;          // مثل IR120170000000101111111001
  shebaDigitsOnly: string;   // صرفاً ۲۴ رقم: 120170000000101111111001
  accountNumber: string;     // شماره حساب استخراج‌شده (مثلاً 101111111001)
  formattedDigits: string;   // فرمت ۲۴ رقمی: 12 0170 0000 0010 1111 1110 01
  formattedSheba: string;    // فرمت کامل: IR12 0170 0000 0010 1111 1110 01
  bank: IranianBankInfo | null;
  errorMessage: string | null;
}

export interface AccountValidationResult {
  isValid: boolean;
  errorMessage: string | null;
  cleanAccountNumber: string;
}

/**
 * اعتبارسنجی سخت‌گیرانه شماره حساب بانکی بر اساس استانداردهای شبکه بانکی کشور
 * (حداقل ۴ رقم، حداکثر ۱۸ رقم، رد قاطع شماره حساب صفر و پوچ)
 */
export function validateAccountNumber(accountNumber?: string | null, bankCode?: string): AccountValidationResult {
  if (!accountNumber || !accountNumber.toString().trim()) {
    return { isValid: false, errorMessage: 'شماره حساب نمی‌تواند خالی باشد.', cleanAccountNumber: '' };
  }
  const cleanDigits = accountNumber.toString()
    .replace(/[۰-۹]/g, d => '۰۱۲۳۴۵۶۷۸۹'.indexOf(d).toString())
    .replace(/[٠-٩]/g, d => '٠١٢٣٤٥٦٧٨٩'.indexOf(d).toString())
    .replace(/\D/g, '');

  if (!cleanDigits) {
    return { isValid: false, errorMessage: 'شماره حساب فاقد هرگونه رقم عددی است.', cleanAccountNumber: '' };
  }
  if (/^0+$/.test(cleanDigits)) {
    return { isValid: false, errorMessage: 'شماره حساب نمی‌تواند صفر یا تماماً صفر باشد.', cleanAccountNumber: cleanDigits };
  }
  if (cleanDigits.length < 4) {
    return { isValid: false, errorMessage: `طول شماره حساب ناقص است (${cleanDigits.length} از حداقل ۴ رقم).`, cleanAccountNumber: cleanDigits };
  }
  if (cleanDigits.length > 18) {
    return { isValid: false, errorMessage: 'طول شماره حساب بیشتر از سقف مجاز بانکی (۱۸ رقم) است.', cleanAccountNumber: cleanDigits };
  }

  return { isValid: true, errorMessage: null, cleanAccountNumber: cleanDigits };
}

/**
 * استخراج شماره حساب بانکی از داخل ارقام شماره شبا (حذف صفرهای زائد سمت چپ)
 */
export function extractAccountNumberFromSheba(input: string): string {
  const digits = extractShebaDigits(input);
  if (digits.length <= 6) return '';
  const rawAccount = digits.substring(6);
  const trimmed = rawAccount.replace(/^0+/, '');
  return trimmed || '';
}

/**
 * یافتن اطلاعات بانک بر اساس نام، نام کوتاه یا کد ۳ رقمی
 */
export function getBankByName(bankName?: string | null): IranianBankInfo | null {
  if (!bankName) return null;
  const clean = bankName.trim();
  return IRANIAN_BANKS.find(b => b.name === clean || b.shortName === clean || b.code === clean || clean.includes(b.name) || b.name.includes(clean)) || null;
}

/**
 * تبدیل مستقیم شماره حساب و بانک به شماره شبای رسمی بر مبنای استاندارد ISO 7064 Mod 97-10 با گارد اعتبارسنجی
 */
export function generateShebaFromAccount(bankIdentifier?: string | null, accountNumber?: string | null, accountType: string = '0'): string {
  if (!bankIdentifier || !accountNumber) return '';
  const bank = getBankByName(bankIdentifier) || IRANIAN_BANKS.find(b => b.code === bankIdentifier);
  if (!bank) return '';

  const accValidation = validateAccountNumber(accountNumber, bank.code);
  if (!accValidation.isValid) return '';

  const paddedAccount = accValidation.cleanAccountNumber.padStart(18, '0');
  const bban = bank.code + (accountType || '0') + paddedAccount;
  const payload = bban + '182700';

  try {
    const remainder = BigInt(payload) % 97n;
    const checkDigits = (98n - remainder).toString().padStart(2, '0');
    return 'IR' + checkDigits + bban;
  } catch {
    return '';
  }
}

/**
 * استخراج هوشمند فقط ارقام عددی از هرگونه رشته، متن، کاراکترهای فارسی، انگلیسی یا نمادها هنگام تایپ و کپی‌پیست
 */
export function extractShebaDigits(input: string): string {
  if (!input) return '';
  // ۱. تبدیل ارقام فارسی و عربی به انگلیسی
  let normalized = input
    .replace(/[۰-۹]/g, d => '۰۱۲۳۴۵۶۷۸۹'.indexOf(d).toString())
    .replace(/[٠-٩]/g, d => '٠١٢٣٤٥٦٧٨٩'.indexOf(d).toString());

  // ۲. استخراج دقیقاً ارقام عددی 0-9 و حذف کلیه حروف، فواصل و نشانه‌ها (مثل I, R, ث, ل, $, _, -, ...)
  const digitsOnly = normalized.replace(/\D/g, '');
  
  // ۳. سقف حداکثر ۲۴ رقم شبا
  return digitsOnly.substring(0, 24);
}

/**
 * پاکسازی و نرمال‌سازی شماره شبا و افزودن قطعی پیشوند IR به ابتدای ۲۴ رقم
 */
export function cleanShebaInput(input: string): string {
  if (!input) return '';
  const digits = extractShebaDigits(input);
  if (!digits) return '';
  return 'IR' + digits;
}

/**
 * فرمت‌بندی ۲۴ رقم عددی به صورت بلوک‌های خوانا (مثلاً: 12 0170 0000 0010 1111 1110 01)
 */
export function formatShebaDigitsDisplay(digits: string): string {
  if (!digits) return '';
  const cleanDigits = digits.replace(/\D/g, '');
  if (cleanDigits.length <= 2) return cleanDigits;
  
  const parts: string[] = [cleanDigits.substring(0, 2)];
  for (let i = 2; i < cleanDigits.length; i += 4) {
    parts.push(cleanDigits.substring(i, i + 4));
  }
  return parts.join(' ');
}

/**
 * فرمت‌بندی نمایشی ۲۶ کاراکتری شبا به صورت بلوک‌های ۴ کاراکتری خوانا
 */
export function formatShebaDisplay(rawSheba: string): string {
  const clean = cleanShebaInput(rawSheba);
  if (!clean || clean.length < 2) return '';
  const digits = clean.substring(2);
  return 'IR ' + formatShebaDigitsDisplay(digits);
}

/**
 * استخراج بانک بر اساس کد ۳ رقمی استاندارد شبا (کاراکترهای ۵ تا ۷ شبا یا ارقام ۳ تا ۵ بدون IR)
 */
export function getBankFromSheba(input: string): IranianBankInfo | null {
  const digits = extractShebaDigits(input);
  if (digits.length < 5) return null;
  // دو رقم اول ارقام کنترلی هستند، ۳ رقم بعدی (اندیس ۲ تا ۵) کد بانک هستند
  const bankCode = digits.substring(2, 5);
  return IRANIAN_BANKS.find(b => b.code === bankCode) || null;
}

/**
 * اعتبارسنجی کامل ریاضی شماره شبا بر مبنای ISO 7064 Mod 97-10
 */
export function validateSheba(input: string): ShebaValidationResult {
  if (!input || !input.trim()) {
    return {
      isValid: false,
      rawSheba: '',
      shebaDigitsOnly: '',
      accountNumber: '',
      formattedDigits: '',
      formattedSheba: '',
      bank: null,
      errorMessage: null
    };
  }

  const digits = extractShebaDigits(input);
  const rawSheba = digits ? 'IR' + digits : '';
  const bank = getBankFromSheba(digits);
  const formattedDigits = formatShebaDigitsDisplay(digits);
  const formattedSheba = formatShebaDisplay(rawSheba);
  const accountNumber = extractAccountNumberFromSheba(digits);

  if (digits.length === 0) {
    return {
      isValid: false,
      rawSheba: '',
      shebaDigitsOnly: '',
      accountNumber: '',
      formattedDigits: '',
      formattedSheba: '',
      bank: null,
      errorMessage: 'شماره شبا شامل هیچ رقمی نیست.'
    };
  }

  if (digits.length < 24) {
    return {
      isValid: false,
      rawSheba,
      shebaDigitsOnly: digits,
      accountNumber,
      formattedDigits,
      formattedSheba,
      bank,
      errorMessage: `طول ارقام شبا ناقص است (${digits.length} از ۲۴ رقم).`
    };
  }

  // الگوریتم رسمی ISO 7064 Mod 97-10
  // ساختار: ۲ رقم اول Check Digits، ۲۲ رقم بعدی شماره حساب
  // چرخش: ۲۲ رقم آخر + 1827 + ۲ رقم اول
  const checkDigits = digits.substring(0, 2);
  if (checkDigits === '00' || checkDigits === '01' || checkDigits === '99') {
    return {
      isValid: false,
      rawSheba,
      shebaDigitsOnly: digits,
      accountNumber,
      formattedDigits,
      formattedSheba,
      bank,
      errorMessage: 'ارقام کنترلی شبا نامعتبر است.'
    };
  }

  // گارد سخت‌گیر: رد شماره حساب‌های تماماً صفر در شماره شبا
  if (!accountNumber || /^0+$/.test(accountNumber)) {
    return {
      isValid: false,
      rawSheba,
      shebaDigitsOnly: digits,
      accountNumber: '',
      formattedDigits,
      formattedSheba,
      bank,
      errorMessage: 'شماره حساب مندرج در شبا نامعتبر (صفر) است.'
    };
  }

  const bban = digits.substring(2); // ۲۲ رقم حساب
  const numericString = bban + '1827' + checkDigits;

  try {
    const remainder = BigInt(numericString) % 97n;
    if (remainder === 1n) {
      return {
        isValid: true,
        rawSheba,
        shebaDigitsOnly: digits,
        accountNumber,
        formattedDigits,
        formattedSheba,
        bank,
        errorMessage: null
      };
    } else {
      return {
        isValid: false,
        rawSheba,
        shebaDigitsOnly: digits,
        accountNumber,
        formattedDigits,
        formattedSheba,
        bank,
        errorMessage: 'کد کنترلی شبا معتبر نیست (خطای محاسباتی Mod 97).'
      };
    }
  } catch {
    return {
      isValid: false,
      rawSheba,
      shebaDigitsOnly: digits,
      accountNumber,
      formattedDigits,
      formattedSheba,
      bank,
      errorMessage: 'خطا در ارزیابی ساختار شبا.'
    };
  }
}
